test/powershell/Language/Scripting/TryCatch.Tests.ps1

```diff
@@ -0,0 +1,651 @@
+        }
+    }
+
+    Context "Mix traps with try/catch" {
+        It "Trap shouldn't catch exception" {
+            $a = . { trap { "test failed" } try { 1; throw "exception"; "test failed" } catch { 2 } }
+            AssertArraysEqual $a (1,2)
+        }
+
+        It "Trap should catch exception" {
+            $a = . { try { 1; throw "exception"; trap { 2; return }; "test failed" } catch { "test failed" } }
+            AssertArraysEqual $a (1,2)
+        }
+    }
+
+    Context "Catch by type" {
+        It "Catch by type #1" {
+            $a = . { try { 1; $a = 0; 1/$a; "test failed" } catch [DivideByZeroException] { 2 } }
+            AssertArraysEqual $a (1,2)
+        }
+
+        It "Catch by type #2" {
+            $a = . { try { 1; $a = 0; 1/$a; "test failed" } catch [DivideByZeroException] { 2 } catch [Exception] { "test failed" } }
+            AssertArraysEqual $a (1,2)
+        }
+
+        It "Catch by type #3" {
+            $a = . { try { 1; $a = 0; 1/$a; "test failed" } catch [DivideByZeroException] { 2 } catch { "test failed" } }
+            AssertArraysEqual $a (1,2)
+        }
+
+        It "Catch by type #4" {
+            $a = . { try { 1; $a = 0; 1/$a; "test failed" } catch [DivideByZeroException],[ArgumentNullException] { 2 } }
+            AssertArraysEqual $a (1,2)
+        }
+    }
+
+    Context "Control flow in try [exit not tested and throw tested elsewhere]" {
+        It "break in try" {
+            $a = . {
+              foreach ($i in (1..3)) {
+                try {
+                  if ($i -eq 2) {
+                    break
+                  }
+                  $i
+                } catch {
+                  "test failed"
+                } finally {
+                  "finally: $i"
+                }
+              }
+            }
+            AssertArraysEqual $a (1, "finally: 1", "finally: 2")
+        }
+
+        It "continue in try" {
+            $a = . {
+              foreach ($i in (1..3)) {
+                try {
+                  if ($i -eq 2) {
+                    continue
+                  }
+                  $i
+                } catch {
+                  "test failed"
+                } finally {
+                  "finally: $i"
+                }
+              }
+            }
+            AssertArraysEqual $a (1, "finally: 1", "finally: 2", 3, "finally: 3")
+        }
+
+        # Disabled - Compiled script has differing (but better) behavior
+        It "return in try" -Skip:$true {
+            $a = . {
+              function foo($i) {
+                try {
+                  if ($i -eq 2) {
+                    return "return: $i"
+                  }
+                  $i
+                } catch {
+                  "test failed"
+                } finally {
+                  "finally: $i"
+                }
+              }
+              foo 1
+              foo 2
+            }
+
+            AssertArraysEqual $a (1, "finally: 1", "finally: 2", "return: 2")
```

Comment:
I've started using `-join ' '` on the array so that I can use `Should Be ...` which somehow feels more idiomatic to Pester even if it's a little weird. This is just a comment to discuss, I'm not asking for a change.